### PR TITLE
Refactoring passing test

### PR DESCRIPTION
**Pull request by orhanna14**
Alright! This is cleaned up quite a bit better I think. I'm still struggling on how to separate the grid data from printing as it stands right now.. could you give me some direction there?

### CODE REVIEW

File: lib/grid.rb [hunk 1]
```diff
@@ -1,24 +1,38 @@
...
+    <<~GRID
+       1  2  3
+       __ __ __
+    A |#{structure["A1"]} |#{structure["A2"]} |#{structure["A3"]} |
+      |__|__|__|
+    B |#{structure["B1"]} |#{structure["B2"]} |#{structure["B3"]} |
+      |__|__|__|
+    C |#{structure["C1"]} |#{structure["C2"]} |#{structure["C3"]} |
+      |__|__|__|
+    GRID
```
Comment (orhanna14) [inline lib/grid.rb]: I'm still not sure how to separate these things with how it's set up. Do you have suggestions?

File: lib/coordinates.rb [hunk 2]
```diff
@@ -0,0 +1,12 @@
+class Coordinates
+  attr_reader :value, :grid
+
+  def initialize(value, grid)
+    @value = value
+    @grid = grid
+  end
+
+  def valid?
+    !grid.coordinate_invalid?(value)
```
Comment (baccigalupi) [inline lib/coordinates.rb]: This expression is pretty hard to follow. Not invalid is a double negative. Does it need to be `coordinate_invalid?`?. If it were checking for valid, then this double negative wouldn't be needed.

File: lib/game_runner.rb [hunk 3]
```diff
@@ -1,37 +1,43 @@
...
 
   def initialize(stdout, stdin, grid = Grid.new)
     @stdout = stdout
     @stdin = stdin
     @grid = grid
-  end
-
-  def printer
-    @printer ||= Printer.new(stdout, grid)
+    @printer ||= Printer.new(stdout)
```
Comment (baccigalupi) [inline lib/game_runner.rb:17]: Since you are setting the `@printer` inside initialize, there is never going to be a situation where it might already exist. So using the `||=` isn't a good approach. Was this copied from somewhere else?

Relatedly, I tend to prefer the instantiation of classes to happen in their own method. That way if something goes wrong in the instantiation, it is easier to debug. If you moved these to methods, then you would use `||=` and you wouldn't have `attr_reader`s for them.

File: lib/game_runner.rb [hunk 4]
```diff
@@ -1,37 +1,43 @@
...
   end
 
   def run
-    printer.print_welcome_message
-    demand_valid_coordinates
+    print_welcome_message
+    print_grid
+    get_valid_coordinate
+    mark_grid
+    print_grid
```
Comment (baccigalupi) [inline lib/game_runner.rb:33]: This method reads really well. That is so important since this is the instruction set for the game. Someone who didn't know the code could very easily understand the app logic by reading this method.

File: lib/game_runner.rb [hunk 5]
```diff
@@ -1,37 +1,43 @@
...
+  end
+
+  def print_grid
+    grid_printer.print_grid
   end
 
-  def coordinate_invalid?
-    !grid.coordinate_valid?(get_user_input)
+  def get_valid_coordinate
+    player_input.get_valid_coordinate
```
Comment (baccigalupi) [inline lib/game_runner.rb:47]: Even though we are getting the coordinate somehow, it isn't returned. That is kind of confusing. I would expect something named `get...` to return a value.

File: lib/game_runner.rb [hunk 6]
```diff
@@ -1,37 +1,43 @@
...
   end
 
-  def get_user_input
-    stdin.gets.chomp
+  def mark_grid
+    player_input.mark_grid
```
Comment (baccigalupi) [inline lib/game_runner.rb]: I don't think the `player_input` should be responsible for marking the grid. Probably this is happening because you are storing the valid coordinate in that player_input object. If instead it was returns and held on to in this class, you could pass it to the grid class and tell the grid to do the marking.

File: lib/grid.rb [hunk 7]
```diff
@@ -1,24 +1,38 @@
...
+  
+  def coordinate_invalid?(coordinate)
+    !structure.key?(coordinate)
+  end
+
+  def get_value(coordinate)
+    structure[coordinate]
+  end
+
+  def current_grid
```
Comment (baccigalupi) [inline lib/grid.rb]: I don't think this is the right name for the method, and I think the naming miss embodies the confusion about where to put this and how to take it apart.

This isn't the current grid. It is a template for the grid, with the logic needed to inject the right values built into the template. That is a ton of logic. You aren't allowed to use external libraries for this solution, but check out the library mustache http://mustache.github.io/mustache.5.html for ideas about logicless templates.

Right now this is a super logic-full template, and you need to convert it to logicless in order to separate the grid structure from the template. That is going to mean that you need to work harder to put the values into the template.

Couple of hints to get you on your way. I would expect another printer type object to ask the grid for its value. It could figure out the coordinates from the template itself. For example, if you had a string template `| {A1} |` you could search this string template for anything matching curly braces with a letter. Pulling that coordinate out of the template itself you could then ask the grid for the value and then substitute it.

I would put this logic inside the grid printer.

File: lib/grid.rb (hunk 1 above)
Comment (baccigalupi) [reply lib/grid.rb]: See comment above!

File: lib/grid_printer.rb [hunk 8]
```diff
@@ -0,0 +1,12 @@
+class GridPrinter
+  attr_reader :stdout, :grid
+
+  def initialize(stdout, grid)
+    @stdout = stdout
+    @grid = grid
+  end
+
+  def print_grid
+    stdout.puts(grid.current_grid)
```
Comment (baccigalupi) [inline lib/grid_printer.rb]: This class seems unnecessary as it is right now. The grid object is doing all the real work of associating data with the template, so that work should come live here.

File: lib/null_coordinates.rb [hunk 9]
```diff
@@ -0,0 +1,9 @@
+class NullCoordinates
```
Comment (baccigalupi) [inline lib/null_coordinates.rb:1]: I LOVE null objects!!

File: lib/player_input.rb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+  def get_coordinate_and_print_error
+    get_coordinate_input
+    unless coordinates.valid?
+      printer.print_coordinates_error
+    end
+  end
+
+  def mark_grid
+    grid.mark(get_valid_coordinate)
```
Comment (baccigalupi) [inline lib/player_input.rb]: I think I already noted that this belongs somewhere else. The name `mark_grid` which is all about another object is the local clue.

File: lib/coordinates.rb [hunk 11]
```diff
@@ -0,0 +1,12 @@
+class Coordinates
+  attr_reader :value, :grid
+
+  def initialize(value, grid)
+    @value = value
+    @grid = grid
+  end
+
+  def valid?
```
Comment (baccigalupi) [inline lib/coordinates.rb:9]: Passing in the grid here where it is needed means that the object identity is not tied to the grid object, but you can use the grid here where you need it.

This pattern of receiving an object in an argument and then querying or commanding that object is called the visitor pattern.

File: spec/coordinates_spec.rb [hunk 12]
```diff
@@ -0,0 +1,87 @@
+require "spec_helper"
+require_relative "../lib/coordinates.rb"
+require_relative "../lib/grid.rb"
+
+RSpec.describe Coordinates do
+  describe "#valid?" do
```
Comment (baccigalupi) [inline spec/coordinates_spec.rb:6]: According to the code, there are only two tests that you need to build:
1. The grid object return false from `invalid_coordinate?`
2. The grid object returns true from `invalid_coordinate?`

The grid object is where all the logic happens so the tests should not be overly familiar with the logic the grid is using to determine that.

File: spec/grid_spec.rb [hunk 13]
```diff
@@ -2,75 +2,14 @@
...
-      valid_row_column = "A1"
+  describe "#mark" do
+    it "takes in a valid coordinate, and gives it a marked value in coordinates" do
+      coordinate = "A3"
       grid = Grid.new
 
-      expect(grid.coordinate_valid?(valid_row_column)).to eq true
-    end
-
-    it "is true if somewhere in the middle of the grid" do
```
Comment (baccigalupi) [inline spec/grid_spec.rb]: So, all of these tests are still important, since this is where the true logic resides.

**Review by baccigalupi — CHANGES_REQUESTED** [review]
Getting better. This is definitely the hardest part of the exercise. Everyone so far has gone a few rounds trying to figure out the right responsibilities.

Let me know if you want to talk tomorrow!

File: lib/game_runner.rb (hunk 3 above)
Comment (orhanna14) [reply lib/game_runner.rb:17]: Thank you! Yes. I was following something from the dependency injection article you had sent me in an older feature, and I don't think I fully understood the reasoning behind it. It makes sense now!

File: lib/game_runner.rb (hunk 6 above)
Comment (orhanna14) [reply lib/game_runner.rb]: Right, I see what you mean. I think I've addressed this in my latest commit. LMK if not!

File: lib/grid.rb (hunk 7 above)
Comment (orhanna14) [reply lib/grid.rb]: Super helpful, and I figured it out! 🙌

File: spec/coordinates_spec.rb (hunk 12 above)
Comment (orhanna14) [reply spec/coordinates_spec.rb:6]: Right! Of course!

File: lib/coordinates.rb (hunk 11 above)
Comment (orhanna14) [reply lib/coordinates.rb:9]: Interesting!
